pkg/sql/subscriber.go

```diff
@@ -246,62 +245,66 @@ func (s *Subscriber) query(
 		"query":      selectQuery,
 		"query_args": sqlArgsToLog(selectQueryArgs),
 	})
-	row := tx.QueryRowContext(ctx, selectQuery, selectQueryArgs...)
-
-	offset, msg, err := s.config.SchemaAdapter.UnmarshalMessage(row)
-	if errors.Cause(err) == sql.ErrNoRows {
-		return "", true, nil
-	} else if err != nil {
-		return "", false, errors.Wrap(err, "could not unmarshal message from query")
+	rows, err := tx.QueryContext(ctx, selectQuery, selectQueryArgs...)
+	if err != nil {
+		return false, errors.Wrap(err, "could not query message")
 	}
 
-	logger = logger.With(watermill.LogFields{
-		"msg_uuid": msg.UUID,
-	})
-	logger.Trace("Received message", nil)
+	var lastOffset int64
+	var lastTransactionID int64
 
-	consumedQuery, consumedArgs := s.config.OffsetsAdapter.ConsumedMessageQuery(
-		topic,
-		offset,
-		s.config.ConsumerGroup,
-		s.consumerIdBytes,
-	)
-	if consumedQuery != "" {
-		logger.Trace("Executing query to confirm message consumed", watermill.LogFields{
-			"query":      consumedQuery,
-			"query_args": sqlArgsToLog(consumedArgs),
+	for rows.Next() {
+		offset, transactionID, msg, err := s.config.SchemaAdapter.UnmarshalMessage(rows)
+		if errors.Cause(err) == sql.ErrNoRows {
+			return true, nil
+		} else if err != nil {
+			return false, errors.Wrap(err, "could not unmarshal message from query")
+		}
+
+		logger = logger.With(watermill.LogFields{
+			"msg_uuid": msg.UUID,
 		})
+		logger.Trace("Received message", nil)
 
-		_, err := tx.ExecContext(ctx, consumedQuery, consumedArgs...)
-		if err != nil {
-			return msg.UUID, false, errors.Wrap(err, "cannot send consumed query")
+		msgCtx := setTxToContext(ctx, tx)
+
+		acked := s.sendMessage(msgCtx, msg, out, logger)
```

Comment:
What do you think, if we send messages concurrently here? As I see the main problem here not in querying batch of messages, but in serial processing of the batch

Reply:
If I'm not missing anything, it would not be compatible with the current backing strategy 🤔 in other words - the current implementation would not work properly with handling out-of-order acks.

Like:

- Msg 1 is acked,
- Msg 2 is acked,
- Msg 3 is not acked
- Msg 4 is acked,

In theory, we could just ack the last message before nack 🤔 but it's a pretty specific scenario. So I think that it may be out of this PR scope. Batch querying already gives a lot of boost compared to querying message by message.

If you'll have some spare time and need such acking model, feel free to post PR with that! (As long as I understood you question properly)